test/symgroups.js

```diff
@@ -0,0 +1,140 @@
+        var type = types.indexOf(sym.group) + 1;
+        tex.write("$" + idx + "+" + key + "+" + type + "\\showlists$\n\n");
+    });
+    tex.end("\\end{document}\n");
+}
+
+function typeset() {
+    var proc = childProcess.spawn(
+        "pdflatex", ["--interaction=nonstopmode", "symgroups"],
+        {stdio: "ignore"});
+    proc.on("exit", function(code, signal) {
+        if (signal) {
+            throw new Error("pdflatex terminated by signal " + signal);
+        }
+        fs.readFile("symgroups.log", "ascii", evaluate);
+    }).on("error", function(err) {
+        throw err;
+    });
+}
+
+/* Consider the symbol "\sim" as an example. At the time of this
+ * writing, it has index 431 in our list, and is of group "rel" which
+ * is the fourth of the types listed above. So we construct an input line
+ * $431+\sim+4\showlists$ and receive corresponding output
+ *
+ * ### math mode entered at line 870
+ * \mathord
+ * .\fam0 4
+ * \mathord
+ * .\fam0 3
+ * \mathord
+ * .\fam0 2
+ * \mathbin
+ * .\fam0 +
+ * \mathrel
+ * .\fam2 '
+ * \mathbin
+ * .\fam0 +
+ * \mathord
+ * .\fam0 4
+ * ### horizontal mode entered at line 870
+ *
+ * This is what we parse, using some regular expressions.
+ */
+
+// Extract individual blocks, from switch to math mode up to switch back.
+var reMM = /^### math mode entered.*\n([^]*?)###/mg;
+
+// Identify the parts separated by the plus signs
+var reParts = /([^]*^\.\\fam0 \+\n)([^]+)(\\mathbin\n\.+\\fam0 \+[^]*)/m;
+
+// Variation of the above in case we have nothing between the plus signs
+var reEmpty = /^\.\\fam0 \+\n\\mathbin\n\.\\fam0 \+/m;
+
+// Match any printed digit in the first or last of these parts
+var reDigit = /^\.\\fam0 ([0-9])/mg;
+
+// Match the atom type, i.e. "\mathrel" in the above example
+var reAtom = /\\([a-z]+)/;
+
+function evaluate(err, log) {
+    if (err) {
+        throw err;
+    }
+
+    var match, nextIndex = 0;
+    while ((match = reMM.exec(log)) !== null) {
+        var list = match[1];
+        match = reParts.exec(list);
+        if (!match) {
+            match = reEmpty.exec(list);
+            if (match) {
+                console.log("Empty: ", nextIndex, keys[nextIndex]);
+                nextIndex++;
+                continue;
+            }
+            console.error("Can't split this into parts:");
+            console.error(list);
+            process.exit(2);
+        }
+        var idx = extractDigits(match[1]);
+        var atom = match[2];
+        var expected = types[extractDigits(match[3]) - 1] || "???";
+        match = reAtom.exec(atom);
+        if (!match) {
+            console.error("Failed to find atom type");
+            console.error(atom);
+            console.error(list);
+            process.exit(3);
+        }
+        var actual = match[1];
+        if (expected !== actual && "math" + expected !== actual &&
+            (expected !== "textord" || actual !== "mathord")) {
+            console.log(expected, actual, idx, keys[idx]);
```

Comment:
It'd be nice if the output included "expected: " + expected + " + "actual: " + actual + ... or something like.